src/fromager/constraints.py

```diff
@@ -12,8 +12,36 @@
 
 
 class Constraints:
-    def __init__(self, data: dict[str, Requirement]):
-        self._data = {canonicalize_name(n): v for n, v in data.items()}
+    def __init__(self) -> None:
+        # mapping of canonical names to requirements
+        # NOTE: Requirement.name is not normalized
+        self._data: dict[NormalizedName, Requirement] = {}
+
+    def __iter__(self) -> typing.Iterable[NormalizedName]:
+        yield from self._data
+
+    def add_constraints(self, *reqs: str) -> None:
+        """Add new constraints, must not conflict with any existing constraints"""
+        for unparsed in reqs:
+            req = Requirement(unparsed)
+            canon_name = canonicalize_name(req.name)
+            previous = self._data.get(canon_name)
+            if previous:
+                # In the future, we could detect non-conflicting requirements
```

Comment:
I'm not sure I would want to merge them. I could definitely see saying that the last (or first) constraint for a package "wins", and others are reported as warnings.

Reply:
Makes sense, I have removed the comment.